src/main/java/net/neoforged/neoforge/transfer/storage/StorageWithSimulations.java

```diff
@@ -0,0 +1,184 @@
+package net.neoforged.neoforge.transfer.storage;
+
+import net.neoforged.neoforge.transfer.Resource;
+
+/**
+ * A generic handler for handling a resource of type {@link T}.
+ * @param <T> The type of resource this handler manages.
+ */
+public interface StorageWithSimulations<T extends Resource> {
+    /**
+     * @return The number of indices this handler manages.
+     */
+    int size();
+
+    /**
+     * Inserts a given amount of the resource into the handler at the given index.
+     *
+     * @param index The index to insert the resource into.
+     * @param resource The resource to insert.
+     * @param maxAmount The amount of the resource to insert.
+     * @param action   The kind of action being performed. {@link TransferAction#SIMULATE} will simulate the action
+     *                 while {@link TransferAction#EXECUTE} will actually perform the action.
+     * @return The amount of the resource that was (or would have been, if simulated) inserted.
+     */
+    int insert(int index, T resource, int maxAmount, TransferAction action);
```

Comment:
Could we use `long` for the return types? Many mods have very large storages for energy/fluids/etc

Reply:
I would prefer it as well.